app/models/manageiq/providers/openstack/inventory/parser/cloud_manager.rb

```diff
@@ -0,0 +1,458 @@
+    end
+  end
+
+  def vms
+    related_infra_ems = collector.manager.provider.try(:infra_ems)
+    hosts = related_infra_ems.try(:hosts)
+
+    collector.servers.each do |s|
+      if hosts
+        # Find associated host from OpenstackInfra
+        filtered_hosts = hosts.select do |h|
+          h.hypervisor_hostname && s.os_ext_srv_attr_host && s.os_ext_srv_attr_host.include?(h.hypervisor_hostname.downcase)
+        end
+        parent_host = filtered_hosts.first
+        parent_cluster = parent_host.try(:ems_cluster)
+      else
+        parent_host = nil
+        parent_cluster = nil
+      end
+
+      availability_zone = s.availability_zone.blank? ? "null_az" : s.availability_zone
+
+      server = persister.vms.find_or_build(s.id.to_s)
+      server[:type] = "ManageIQ::Providers::Openstack::CloudManager::Vm"
+      server[:uid_ems] = s.id
+      server[:ems_ref] = s.id
+      server[:name] = s.name
+      server[:vendor] = "openstack"
+      server[:raw_power_state] = s.state || "UNKNOWN"
+      server[:connection_state] = "connected"
+      server[:location] = "unknown"
+      server[:host] = parent_host
+      server[:ems_cluster] = parent_cluster
+      server[:availability_zone] = persister.availability_zones.lazy_find(availability_zone)
+      server[:key_pairs] = [persister.key_pairs.lazy_find(s.key_name)].compact
+      server[:cloud_tenant] = persister.cloud_tenants.lazy_find(s.tenant_id.to_s)
+      server[:parent] = persister.miq_templates.lazy_find(s.image["id"]) unless s.image["id"].nil?
+
+      # to populate the hardware, we need some fields from the flavor object
+      # that we don't already have from the flavor field on the server details
+      # returned from the openstack api. It's possible that no such flavor was found
+      # due to some intermittent network issue or etc, so we use try to not break.
+      flavor = collector.flavors_by_id[s.flavor["id"].to_s]
+      if flavor.nil?
+        # the flavor might be private, which the flavor list doesn't seem to handle
+        # correctly. fetch it now and build an inventory object for it.
+        flavor = collector.private_flavor(s.flavor["id"].to_s)
+        make_flavor(flavor) unless flavor.nil?
+      end
+      server[:flavor] = persister.flavors.lazy_find(s.flavor["id"].to_s)
+      server[:hardware] = persister.hardwares.find_or_build(s.id)
+      server[:hardware][:vm_or_template] = persister.vms.lazy_find(s.id)
+      server[:hardware][:cpu_sockets] = flavor.try(:vcpus)
+      server[:hardware][:cpu_total_cores] = flavor.try(:vcpus)
+      server[:hardware][:cpu_speed] = parent_host.try(:hardware).try(:cpu_speed)
+      server[:hardware][:memory_mb] = flavor.try(:ram)
+      server[:hardware][:disk_capacity] = (
+        flavor.try(:disk).to_i.gigabytes + flavor.try(:swap).to_i.megabytes + flavor.try(:ephemeral).to_i.gigabytes
+      )
+
+      server[:hardware][:networks] = []
```

Comment:
I think we don't need this? And the assignment. Since network can save on it's own, providing the :hardware. Saving through association is possible too, but it's slower in general. I use saving through association usually for simple N:M mapping tables